### DISCUSSION TITLE
return custom class with multiple hasmany

### DISCUSSION THREAD

**Discussion by rizalahmaddd** [simolus3/drift#2099] [Q&A]
i have custom class like this

<img width="331" alt="Screenshot 2022-10-12 at 8 40 33 AM" src="https://user-images.githubusercontent.com/93845700/195247636-163caf0f-dead-4270-96fb-6f1c7827c575.png">

and this is the query
<img width="564" alt="Screenshot 2022-10-12 at 8 40 48 AM" src="https://user-images.githubusercontent.com/93845700/195247683-d0d4daed-bfd0-4e97-86cc-c78f0e08a2bb.png">

occure error, i dont know how to return properly. 
<img width="1219" alt="Screenshot 2022-10-12 at 11 14 45 AM" src="https://user-images.githubusercontent.com/93845700/195248560-97e53968-49d0-4953-8137-2462fd6b0bdb.png">

thanks in advance.

**Comment by simolus3** [comment]
In a single SQL query, you only get one joined table per row. You can't have "list" results in a single row in SQL. The only way to fix this is to rewrite the query, but it gets a lot more complicated.

Essentially, you can follow [this guide](https://drift.simonbinder.eu/docs/examples/relationships/) and adapt it to your use case with two nested results. You can slightly simplify things because you don't need to watch tables. Essentially, it boils down to:

1. Running the query without `surveyOptions` and `surveyAnswerDetails`.
2. Constructing a list of every `surveyQuestion.id`
3. Running a new query, only on `surveyOptions` and `surveyAnswerDetails`. Filter for ids contained in the list created in step 2.
4. Use lookups in Dart to stitch the results together.

If you're not afraid to write direct SQL, you can let drift take care of this transformation for you. You can add a query like this to the `@DriftDatabase` annotation:

```dart
@DriftDatabase(
  ....,
  queries: 'questionWithAnswerDetail': '''
SELECT 
  survey_questions.**,
  survey_question_types.**,
  LIST(SELECT * FROM survey_answer_details WHERE survey_answer_id = :answerId AND survey_question_id = survey_questions.id) AS answers,
  LIST(SELECT * FROM survey_options WHERE survey_question_id = survey_questions.id) AS options
FROM survey_questions
  LEFT OUTER JOIN survey_question_types ON survey_question_types.id = survey_questions.question_type_id,
''',
)
```

(Might have some syntax errors).

Here:

- The `.**` tells drift to [include the result as a structured row class](https://drift.simonbinder.eu/docs/using-sql/drift_files/#nested-results).
- The `LIST` is a drift macro. The compiler will [run multiple queries for you to emulate this](https://drift.simonbinder.eu/docs/using-sql/drift_files/#list-subqueries).
